keps/sig-node/4188-kubelet-pod-readiness-api/README.md

```diff
@@ -0,0 +1,450 @@
+}
+
+// PodCondition aligns with v1.PodCondition.
+message PodCondition {
+    PodConditionType Type = 1;
+    ConditionStatus Status = 2;
+    Timestamp LastProbeTime = 3;
+    Timestamp LastTransitionTime = 4;
+    string Reason = 5;
+    string Message = 6;
+}
+
+// PodConditionType aligns with v1.PodConditionType
+enum PodConditionType {
+    ContainersReady = 0;
+	Initialized = 1;
+	Ready = 2;
+	PodScheduled = 3;
+	DisruptionTarget = 4;
+}
+
+// ConditionStatus aligns with v1.ConditionStatus
+enum ConditionStatus {
+    True = 0;
+	False = 1;
+	Unknown = 2;
+}
+
+// PodInfo returns a Pod details and list of status Conditions with deletion info.
+message PodInfo {
+    string podUID = 1;
+    string podNamespace = 2;
+    string podName = 3;
+    bool static = 4;
+	repeated PodCondition conditions = 5;
+    Timestamp DeletionTimestamp = 3;
+}
+
+// PodConditionsResponse returns a stream of List of PodConditionsResponse
+// Whenever a Pod state changes it will return the new list.
+message PodConditionsListResponse {
+	repeated PodInfo Pods = 1;
+}
+
+// PodConditionsRequest contains Pods UID
+message PodConditionsRequest {
+    string podUID = 1;
+}
+```
+
+##### Unit tests
+
+
+##### e2e tests
+
+  - Feature gate enable/disable tests.
+  - Test API work when Kubelet restarts.
+  - Test API work when Control Plane is unreachable.
+  - Test API to work with kubelet Standalone mode.
+
+### Graduation Criteria
+
+#### Alpha
+
+- [ ] Feature implemented behind a feature flag.
+- [ ] e2e tests completed and enabled.
+- [ ] sig-auth input on proposed lack of authorization for this API.
+
+#### Beta
+- [ ] Fix Kubelet restart [issue](https://github.com/kubernetes/kubernetes/issues/100277)
+
+#### GA
+
+- [ ] No major bugs related to usage of API in 3 months cycle.
+
+### Upgrade / Downgrade Strategy
+
+With gRPC the version is part of the service name.
+Old versions and new versions should always be served and listened by the Kubelet.
+
+### Version Skew Strategy
+
+N/A
+
+## Production Readiness Review Questionnaire
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: `PodInfoAPI` is a new feature gate to
+  enable / disable PodInfo API.
+  - Components depending on the feature gate: Kubelet
+
+###### Does enabling the feature change any default behavior?
+
+No.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, through feature gates.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+The API becomes available again. The API is stateless, so no recovery is needed, clients can just consume the data.
+
+###### Are there any tests for feature enablement/disablement?
+
+e2e test will demonstrate that when the feature gate is disabled, the API returns the appropriate error code.
```

Comment:
This will effectively be a feature test.

Given this is purely in-memory feature - enablement/disablement is effectively equal to the feature test though [but let's clarify that in the text here].